### ISSUE TITLE
Use a different word for highlight

### ISSUE THREAD

**Issue by csillag** [hypothesis/h#2054]
This came out from the discussion [here](https://github.com/hypothesis/h/pull/2036#issuecomment-82608733).

@tilgovi says:

> From now on I'm not going to call something a highlight unless it's highlighted. 

So we need a proper name for the *things* that we used to call *highlights* (which are, in most cases, SPAN elements around pieces of text), and use that name everywhere in our code andUI.

I have no preference, except that I really think we should avoid the words that are already used for other things, including target, anchor and annotation.

How about maybe "mark" ?

**Comment by csillag** [comment]
@tilgovi, would you be comfortable with calling the SPANs that we add to the DOM (which are sometimes visible, sometimes not) *marks* ?

**Comment by csillag** [comment]
Also, in the same run, we can do the following changes:
 * *activate* (a highlight) -> *highlight* (a mark)
 * *deactivate* (a highlight) -> *unhighlight* or *remove highlight* ?
 * *show annotations* -> *highlight all marks*

I believe that doing these renamings would achieve what @tilgovi wanted - that is, never call anything a highlight which is not actually highlighted.

**Comment by tilgovi** [comment]
I think `target` is fine. The target may or may not exist in the fetched document representation even if it is described in the annotation. But the concept is the same. I mean, there's a little fuzziness in the term there due to the constraints of the span-based approach (i.e. there is not necessarily one span per target) but logically I think it's 1:1 still, for the purposes of human reasoning and discussion.

**Comment by tilgovi** [comment]
:-1: for "mark" for the same reason as "highlight": to a human it sounds like a visible thing.

**Comment by tilgovi** [comment]
Also, let's bear in mind for this discussion that this language would not be user facing. We're not telling people to "make targets" or anything like that. Users make highlights.

**Comment by tilgovi** [comment]
"Target" makes a ton of sense for me. It's also the term of art for the locus of an event in DOM specifications. Targets are usually `Element`s or, more generally, `Node`s. Our spans are there for highlighting for also as an interaction point, a.k.a "target".

**Comment by csillag** [comment]
I think that overloading target with a new meaning just invites 
confusion.

In our code, we already have things that we call targets: the data 
structures that we use to describe the segment of the document that is 
the subject of the annotation. They consist of a list selectors.

What we are talking about now is things that live in the DOM.

That's just not the same thing.

Also, many of our data structures (annotations, anchors, etc) have 
properties and methods that deal with targets and highlights. Starting 
to use the same word for those would be very confusing.

**Comment by tilgovi** [comment]
I'm suggesting that it isn't a new meaning. It's not overloading because it's the same concept. A thing with a location in a document model.

**Comment by tilgovi** [comment]
Or rather, the location itself.

**Comment by tilgovi** [comment]
Your name is your name whether spoken or written. You could specify, but there's an umbrella type/term and it's "name". A target is a target whether serialized in an annotation, a `Node` in a tree or a point on your screen.

**Comment by csillag** [comment]
> I'm suggesting that it isn't a new meaning. It's not overloading because it's the same concept.
> A thing with a location in a document model. 
> [...]
> Or rather, the location itself.

> Your name is your name whether spoken or written. You could specify, but there's an umbrella
> type/term and it's "name". A target is a target whether serialized in an annotation,
>  a `Node` in a tree or a point on your screen.

Using umbrella terms is perfectly fine when we are just mentioning something occasionally.

However, in our discussions, these concepts are often at the center of attention. Therefore, we *need* to be able to distinguish between the different forms.

I guess you wouldn't like to use the same term for "user", "account", and "token", even if all of those are basically talking about who is using our system. In the auth code, you need different names for all those concepts, and when discussing the auth code, you will use those names.

I patently reject to confuse our discussions by starting to refer to technically very different concepts with the same noun, at least in the discussions which are about the very topics.

**Comment by tilgovi** [comment]
I think your comparison is poor. The `user` and `account` I would use interchangeably, but `token` I would not because it may or may not bear any authorization for one or more `user`, or more accurately `principal`.

But I'm not here to argue about the examples. I suggest we use the term `target` going forward and _if it becomes an issue_, as in conversation is routinely confusing and it's ambiguous which thing is being referred to, then we should open a new issue.

**Comment by tilgovi** [comment]
Keep in mind "target" _already has a meaning_ in the DOM domain and it is very compatible, I think, with how I'm using it to refer to spans or other node constructions we may highlight using CSS.

**Comment by tilgovi** [comment]
Sorry, I want to be clear that I'm not trying to cut this short. If you think the target language is a real issue, we can change this issue to reflect that or start a new one. However, I don't think we know until we try. I'm proposing we try calling them targets. I do think referring to things that aren't actually highlighted as "highlights" is a bit confusing.

**Comment by csillag** [comment]
So, here is a very short summary of some of our relevant terms, as we use them today:
 * `target`: the part of the document we are annotating
 * `selectors`: the chunks of information we are using to describe `targets`
 * `anchors`: our *best guess* about where the *target* should be in the current version of the document.
 * `highlights`: the things we put into the DOM, to visually indicate to the user where our *anchors* are.

You are saying that we should rename `highlights` to `targets`, because they are really the same thing: just locations inside the document

My objections are:
 * First: actually, it's not the *highlight*, but the *anchor* concept that is more closely related to the `target`, because they are both locations inside the document. (`target` is the actual location during the creation of the annotation; `anchor` is the location calculated/guessed by us.) In contrast, *highlights* are merely *visual indicators* of the location. Therefore, renaming *highlights* to *targets*, while the more closely related *anchor* would have a different name, would be very misleading. (Of course, renaming all three things to the same would be even more misleading, so I'm not suggesting that.)
 * Secondly, there are parts of our code that are referencing `targets`, `anchors` and `highlights` *within the same methods*. (See, for example, various parts of [the Anchor class](https://github.com/hypothesis/h/blob/034cb4e15c9898196aed6a7029d4af7cf3d46e71/h/static/scripts/annotator/plugin/enhancedanchoring.coffee#L15), which has a reference to the `target` it was created for, and also has a list of the `highlights` that we are using for indicating the presence of that `anchor` to the users.) Renaming any of those concepts would involve renaming those fields, which would mean that we would end up having the same name in our code for things that are clearly different. The other option would be to leave the code alone, and use these names only in our communication, but using different wording for discussing and writing the code would cripple our discussions even more.
 * Thirdly, we are doing all this renaming in order to clean up existing confusion caused by using the same word for two different things. I see no reason whatsoever to fall into the same pit again.

> If you think the target language is a real issue,

It would be if we used it that way. Fortunately, we don't.

> we can change this issue to reflect that or start a new one. 

Fortunately, that won't be necessary, since don't have that problem in our code / UI / communications, and hopefully, we never will.

> However, I don't think we know until we try.

Oh yes we do. I am pretty sure it would be obvious to everyone right at the instant when
 - someone tried to draw up a diagram about how the different concepts / classes / object of our code relate to each other
 - someone would take a glance at our "improved" code, in which we have methods that use the same variable or field names in the same scope for totally different objects.

In short: :-1: :-1: :-1: :-1: for using the `target` word for what we currently call `highlight`. 

> I do think referring to things that aren't actually highlighted as "highlights" is a bit confusing.

And I agree with that! So let's find a different word, but please pick a one that would not recreate the confusion in an other area.

   * * *

So we are searching for a word that mean something that *can* be visible, but isn't necessarily visible all the time. Something that can be switched on and off. My new proposal is:
 * `indicator`. Taking the meaning for chemistry, and indicator can be present in a solution without being visible; however it will change it's color when something interesting happens.

**Comment by tilgovi** [comment]
Anchors are `<a>` tags.

Targets are the locus points of interactions, such as annotations, clicks, etc.

Anchors, as you use them, is something you made up by yourself that has never seemed justified to me and I've argued against it repeatedly.

You've also got 'segment', which is superfluous as well as far as I'm concerned.

**Comment by csillag** [comment]
Let's do an experiment. Check out this (hypothetical) explanation that we would get if we renamed `highlights` to `targets`:

> We start out with the target. We calculate an anchor from the data available in the target. Then immediately after that, we create a target for this anchor.

This would be supremely confusing, since
 * The whole workflow starts out with a target; why would we create the target at the end of the process?
 * The two targets in this section are (technically and conceptually) completely different things; the first one is `a collection of information` about a location, and the second one is an `indicator` for a location.

**Comment by tilgovi** [comment]
Just state it more simply:

    We start out with a serialized target and then find or create an appropriate representation of it in the DOM.

**Comment by tilgovi** [comment]
We're just changing a `String` to a `Node`. Nothing more complicated.

**Comment by csillag** [comment]
> Anchors, as you use them, is something you made up by yourself that has never seemed justified to me and I've argued against it repeatedly.

As far as I can remember, you haven't objected to it when it was introduced, nor do i remember seeing a proposal to rename it to something else later on. Anyway, I am fine with widening the scope of the discussion to cover that term, too.

So, what do you propose we should call what we currently call `anchors` ?

> You've also got 'segment', which is superfluous as well as far as I'm concerned.

As far as I can tell, it only appears at 1 or 2 places of the scope, (as opposed to target, anchor and highlight, which are often used), so I don't think it's significant. But we can get to it, too.

**Comment by csillag** [comment]
> Anchors are <a> tags.

That's not the only meaning.
Take, for example, the term *robust anchoring*. Is it about <a> tags? No. Then what is it about? It's about clinging to places. In an extended sense, and anchor is a method for clinging to places.

We are using the word in that sense.

I agree that it's not ideal that HTML has a conflicting term; I am happy to pick another. Just pick something that doesn't conflict with an existing term, please.

**Comment by tilgovi** [comment]
How about "target"?

**Comment by csillag** [comment]
> Just state it more simply:

> > We start out with a serialized target and then find or create an appropriate representation of it in the DOM.

This proposed explanation doesn't cover the same ground, because it's missing an important interim step: before we create anything in the DOM, first we identify the wanted place in the DOM. We need a term for that, too. 

Furthermore, you are mentioning *a representation* (of the target). A representation of something is not the same thing as the the object that is being represented. It is perfectly normal that it has a different name. And the *highlight* is the name of the representation of that location.

**Comment by tilgovi** [comment]
All the things are representations.

**Comment by csillag** [comment]
Still, we use different words for
 * money
 * bank account
 * credit card
 * cash

.. even though they still represent some exchangeable value, but in a different shape or form.
Just like we do here.

**Comment by tilgovi** [comment]
http://www.w3.org/TR/html401/struct/links.html#h-12.1

Links have anchors at both ends. Since an annotation describes a relationship, a link, between two anchors, anchor is not a bad term.

"Anchor" is fine. But it's just a matter of perspective. The place in the document is an "anchor" with respect to the link, but it's a a "target" with respect to the annotation / act of annotating.

**Comment by tilgovi** [comment]
In other words, if you wanted to call them anchors I think you have a sound argument.

**Comment by csillag** [comment]
> In other words, if you wanted to call them anchors I think you have a sound argument.

Who is "them" in this sentence? Because we are now discussing renaming `anchors` and `highlights`.

**Comment by tilgovi** [comment]
I am _not_ discussing highlights. It is totally irrelevant whether one decides to highlight the target. That's how we got to this issue to begin with.

**Comment by csillag** [comment]
> I am not discussing highlights.

This issue is about finding a new name for the things that we used to call highlights.
If you are not discussing that, then I no longer understand what you are talking about.

> It is totally irrelevant whether one decides to highlight the target.

Hey, wait a minute. In the above sentence, you are using *highlight* as an action, which is *not* the meaning this issue is about. 

This issue is about renaming *the other* usage (which is a *thing*, not an *action*), so that the action will be the only remaining meaning.

**Comment by tilgovi** [comment]
Action: "to highlight".
Thing: "a highlight".

It is both an action and a thing. The issue is about using it to only mean one kind of thing rather than two.

You made a good case that anchor is a fine word, if you want to take the link-centric view of the thing. I argue that if you're taking an interaction-centric view "target" is probably better.

I only say "I am not discussing highlights" because I made it clear already that I don't want to refer to these things as highlights. I am discussing a thing, but it is not a highlight. We're discussing the name for that thing, as you correctly state. Since we're on the same page about that, I hoped that we could agree we aren't discussing highlights.

**Comment by csillag** [comment]
> It is both an action and a thing. The issue is about using it to only mean one kind of thing rather than two.

Exactly.

>  I am discussing a thing, but it is not a highlight. 

Well, technically, I don't think that your desire (to call it something else) already had the effect of changing the name of the thing; I would say that the name will change in the instant when we come up with a new name and make a decision about that.

But I'm not in the mood for semantic games, so for the scope of this discussion, I already started to call them "the things that we used to call highlights".

Let's stick with that for this issue, shall we? TTTWUTCH - "the things that we used to call highlights".

So, how should we call TTTWUTCH ?

My current proposal is "indicator", because it kind of captures that is there to indicate (or represent, as you said) something else. We could also call it XXXIndicator to make the association between XXX and the XXXIndicator more obvious. (XXX could be either Anchor or Target)

Along the same lines, I am open to rename `Anchor` to `TargetLocation`, and `Target` to `TargetDescription`. This would result:
  * `TargetDescription` is the serialized form, where the target is described with selectors
  * `TargetLocation` is the location that we calculated for the target, looking at the selectors and the document
  * `TargetIndicator` is what we place in the DOM to (sometimes) indicate this location to the users/

How does that sound? This would express that ultimately, these are all related to the target, but they would still make it possible for us to have a meaningful conversation about them, without being forced to use the very same word to describe them.

**Comment by tilgovi** [comment]
If you wanted to use those as variable names those would make sense. I'd
hope for camel case because otherwise (and this is the current situation)
this is excessively object oriented. That's really what I object to at the
heart of all this: what we're taking about is various stages of
transformation of the same data, but not really different types and there
are no behaviors.

The agent finds a target, anchors a target, highlights a target,
deserializes or serializes a target. The user selects a target. Etc.

So, sure. Talking about the location of a target, or its selectors, or
whatever, all makes sense.

**Comment by tilgovi** [comment]
But what I'm trying to say is we should get out of this object oriented
world view. If you do that, I suspect each function would have a clear
context in which the meaning of "target" is singular and clear. There's no
need for any other terms.

**Comment by nickstenning** [comment]
Can I suggest that this is a conversation better suited to a mailing list thread than a GitHub issue?

**Comment by tilgovi** [comment]
Yes.
